File: sdk/python/symbolchain/core/nem/KeyPair.py
import sha3

from ..CryptoTypes import PrivateKey, PublicKey, Signature
from .external import ed25519


class KeyPair:
    """Represents an ED25519 private and public key."""

    def __init__(self, private_key):
        """Creates a key pair from a private key."""
        self._sk = private_key.bytes[::-1]
        self._pk = ed25519.publickey_hash_unsafe(self._sk, sha3.keccak_512)

    @property
    def public_key(self):
        """Gets the public key."""
        return PublicKey(self._pk)

    @property
    def private_key(self):
        """Gets the private key."""
        return PrivateKey(self._sk[::-1])

    def sign(self, message):
        """Signs a message with the private key."""
        return Signature(ed25519.signature_hash_unsafe(message, self._sk, self._pk, sha3.keccak_512))


class Verifier:
    """Verifies signatures signed by a single key pair."""

    def __init__(self, public_key):
        """Creates a verifier from a public key."""
        self._pk = public_key.bytes

    def verify(self, message, signature):
        """Verifies a message signature."""
        try:
            ed25519.checkvalid_hash(signature.bytes, message, self._pk, sha3.keccak_512)
            return True
        except (ValueError, ed25519.SignatureMismatch):
            return False

File: sdk/javascript/src/nem/KeyPair.js
const { PrivateKey, PublicKey, Signature } = require('../CryptoTypes');
const ed25519 = require('../impl/ed25519');
const { deepCompare } = require('../utils/arrayHelpers');

const HASH_MODE = 'Keccak';

/**
 * Represents an ED25519 private and public key.
 */
class KeyPair {
	/**
	 * Creates a key pair from a private key.
	 * @param {PrivateKey} privateKey Private key.
	 */
	constructor(privateKey) {
		this._privateKey = privateKey;

		const reversedPrivateKeyBytes = new Uint8Array([...privateKey.bytes]);
		reversedPrivateKeyBytes.reverse();
		this._keyPair = ed25519.keyPairFromSeed(HASH_MODE, reversedPrivateKeyBytes);
	}

	/**
	 * Gets the public key.
	 * @returns {PublicKey} Public key.
	 */
	get publicKey() {
		return new PublicKey(this._keyPair.publicKey);
	}

	/**
	 * Gets the private key.
	 * @returns {PrivateKey} Private key.
	 */
	get privateKey() {
		return new PrivateKey(this._privateKey.bytes);
	}

	/**
	 * Signs a message with the private key.
	 * @param {Uint8Array} message Message to sign.
	 * @returns {Signature} Message signature.
	 */
	sign(message) {
		return new Signature(ed25519.sign(HASH_MODE, message, this._keyPair.privateKey));
	}
}

/**
 * Verifies signatures signed by a single key pair.
 */
class Verifier {
	/**
	 * Creates a verifier from a public key.
	 * @param {PublicKey} publicKey Public key.
	 */
	constructor(publicKey) {
		if (0 === deepCompare(new Uint8Array(PublicKey.SIZE), publicKey.bytes))
			throw new Error('public key cannot be zero');

		this.publicKey = publicKey;
	}

	/**
	 * Verifies a message signature.
	 * @param {Uint8Array} message Message to verify.
	 * @param {Signature} signature Signature to verify.
	 * @returns {boolean} true if the message signature verifies.
	 */
	verify(message, signature) {
		return ed25519.verify(HASH_MODE, message, signature.bytes, this.publicKey.bytes);
	}
}

module.exports = { KeyPair, Verifier };
